File: src/java-sample/com/xtaudio/xt/CaptureSimple.java
package com.xtaudio.xt;

import java.io.FileOutputStream;

public class CaptureSimple {

    static final int SAMPLE_SIZE = 3;
    static final XtFormat FORMAT = new XtFormat(new XtMix(44100, XtSample.INT24), 1, 0, 0, 0);

    static void capture(XtStream stream, Object input, Object output, int frames, double time,
            long position, boolean timeValid, long error, Object user) throws Exception {

        if (frames > 0)
            // Don't do this.
            ((FileOutputStream) user).write((byte[]) input, 0, frames * SAMPLE_SIZE);
    }

    public static void main(String[] args) throws Exception {

        try (XtAudio audio = new XtAudio(null, null, null, null)) {
            XtService service = XtAudio.getServiceBySetup(XtSetup.CONSUMER_AUDIO);
            try (XtDevice device = service.openDefaultDevice(false)) {
                if (device != null && device.supportsFormat(FORMAT)) {

                    XtBuffer buffer = device.getBuffer(FORMAT);
                    try (FileOutputStream recording = new FileOutputStream("xt-audio.raw");
                            XtStream stream = device.openStream(FORMAT, true, false,
                                    buffer.current, CaptureSimple::capture, null, recording)) {
                        stream.start();
                        Thread.sleep(1000);
                        stream.stop();
                    }
                }
            }
        }
    }
}

File: src/cpp-sample/capture-simple.cpp
#include <xt-cpp.hpp>
#ifdef _WIN32
#include <windows.h>
#else
#include <unistd.h>
#endif
#include <fstream>
#include <iostream>
#include <cmath>

static const int SampleSize = 3;
static const Xt::Format Format(Xt::Mix(44100, Xt::Sample::Int24), Xt::Channels(1, 0, 0, 0));

static void Capture(
  const Xt::Stream& stream, const void* input, void* output, int32_t frames,
  double time, uint64_t position, bool timeValid, uint64_t error, void* user) {

  if(frames > 0)
    // Don't do this.
    static_cast<std::ofstream*>(user)->write(static_cast<const char*>(input), frames * SampleSize);
}

int CaptureSimpleMain(int argc, char** argv) {

  Xt::Audio audio("", nullptr, nullptr);  
  auto system = Xt::Audio::SetupToSystem(Xt::Setup::ConsumerAudio);
  std::unique_ptr<Xt::Service> service = Xt::Audio::GetService(system);
  if(!service)
    return 0;  

  std::unique_ptr<Xt::Device> device = service->OpenDefaultDevice(false);
  if(!device || !device->SupportsFormat(Format))
    return 0;

  Xt::Buffer buffer = device->GetBuffer(Format);
  std::ofstream recording("xt-audio.raw", std::ios::out | std::ios::binary);
  std::unique_ptr<Xt::Stream> stream = device->OpenStream(Format, true, buffer.current, Capture, nullptr, &recording);
  stream->Start();
#if _WIN32
  Sleep(1000);
#else
  usleep(1000 * 1000);
#endif
  stream->Stop();
  return 0;
}